### PR TITLE
Add parameter to cite allowing for user-defined styling (#1220)

### PR DESCRIPTION
**Pull request by Gutawer**
This draft implements #1220. The basic functionality to allow for user-defined styling is implemented via changing `Works` to instead store either `Content`, or a function call with arguments that the element will call when it needs to `Show`.

The dictionary provided to user code is mostly identical to `hayagriva`'s public API, besides a few points - the first is that `FmtString` isn't represented, since it only consists of one non-private member. The second is that `affiliated_persons` is represented with a dictionary instead of a tuple. The third is that urls are currently just exposed as a string rather than the parsed data structure. All of these can be more-or-less easily changed, of course.

Probably the most contentious part of the PR as it stands is that the handling of the conversion from `hayagriva` data to `typst` values is manual and verbose, and also might be better off in `hayagriva` itself.

### CODE REVIEW

**Comment by Gutawer** [comment]
Here's some example usage of the API:
```
#let citeauthor(key, min: 1, max: 3) = cite(key, style: dict => {
    let authors = dict.authors.map(a => a.name)
    let trunc = authors.len() > max
    if trunc {
        authors = authors.slice(0, min)
    }
    if trunc {
        authors.join(", ") + [ _et al._]
    } else {
        (
            authors.slice(0, authors.len() - 1).join(", ") +
            [ and ] +
            authors.at(authors.len() - 1)
        )
    }
})

#let citeyear(key) = cite(key, style: dict => {
    [#dict.date.year]
})
```
This would allow for functionality similar to `biblatex`' `citeauthors` and `citeyear` as a stop-gap measure before they are added to typst in a more built-in way.

**Comment by PgBiel** [comment]
I believe this will be easier (involve less manual work to convert to dict and stuff) and more accurate if Hayagriva exposes the `contents` field of `Entry` directly, so that any specified properties in the original entry are given, and also as it's already a HashMap (so conversion to a dict would be likely simpler). @laurmaedje thoughts? (Would my suggestion be better or worse than the current approach?)

**Comment by laurmaedje** [comment]
Exposing the HashMap seems okay to me, maybe as a `fields() -> &HashMap<..>` function. Note that when converting the hash map to a Typst dict, you need to take care to produce a dictionary with predictable order. Rust's HashMap iteration order is random and we can't have randomness in Typst.

**Comment by PgBiel** [comment]
Thanks for the input. We can try playing around with this locally then, to perhaps prepare for a hayagriva PR.

Though, I wonder if hayagriva _itself_ could use an IndexMap (or whatever it's called) then? Would that be too bad? I think that would be a more sane solution than just pinning an arbitrary order here, haha (not that it would matter that much though).

**Comment by laurmaedje** [comment]
Perhaps, although the biblatex -> hayagriva conversion probably wouldn't preserve order.

**Comment by PgBiel** [comment]
> Perhaps, although the biblatex -> hayagriva conversion probably wouldn't preserve order.

Ah, that's interesting. Then maybe we should just sort the HashMap keys alphabetically on the Typst side and be done with it

**Comment by laurmaedje** [comment]
@Gutawer What's the state of this?

**Comment by Gutawer** [comment]
Well, provided the approach of manually exposing the hayagriva API is OK (and personally I feel like it is cleanest to expose what was already chosen as good to expose in hayagriva?), it should be good to merge once the following things are checked:

- The typst representation of the data changes `affiliated_persons`, which in hayagriva has type `Option<&[(Vec<Person>, PersonRole)]>`, to instead use a dictionary instead of a tuple - I did this because using an array-of-array-of-array structure doesn't sound particularly nice to work with in the dynamically-typed typst context.

- The full url-crate data structure for URLs is not exposed currently, instead sticking the string representation into a dictionary entry called `value-string`. This should be good for most use cases where making the URL string into content is what is intended.

- **Most importantly**, the PR doesn't represent `FmtString`s in any way besides turning their `value` into a typst string. This is not the most flexible behaviour possible, and I'm now realizing that it's not really future-compatible - I'd like advice on this because I'm not totally aware on the context in which the formatting functions would be called, and subsequently how would be best to fix this.

Other than this, the design is just the hayagriva API, converted to typst values.

File: crates/typst-library/src/meta/bibliography.rs [hunk 1]
```diff
@@ -445,6 +481,144 @@ impl Works {
...
+
+/// Converts a hayagriva date (or something which can borrow as one) into a typst dictionary.
+fn date_to_dict(date: impl std::borrow::Borrow<hayagriva::types::Date>) -> Dict {
+    let date = date.borrow();
+    dict! {
+        "year" => date.year,
+        "month" => date.month,
+        "day" => date.day,
+    }
+}
```
Comment (PgBiel) [inline crates/typst-library/src/meta/bibliography.rs:539]: Shouldn't we use typst's `datetime` type here?

Comment (Gutawer) [reply crates/typst-library/src/meta/bibliography.rs:539]: Definitely, I did this before that was in typst IIRC.

Comment (PgBiel) [reply crates/typst-library/src/meta/bibliography.rs:539]: Yeah, makes sense haha. This is looking pretty good either way. Take your time!

File: crates/typst-library/src/meta/bibliography.rs [hunk 2]
```diff
@@ -445,6 +481,144 @@ impl Works {
...
+
+/// Converts a hayagriva qualified url into a typst dictionary.
+///
+/// For now, this doesn't expose the full `url` crate type, instead just providing `value` as a
+/// string.
+fn qualified_url_to_dict(url: &hayagriva::types::QualifiedUrl) -> Dict {
+    dict! {
+        "value-string" => url.value.to_string(),
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/bibliography.rs:581]: The name `value-string` is a bit strange in my opinion.

File: crates/typst-library/src/meta/bibliography.rs [hunk 3]
```diff
@@ -391,9 +397,9 @@ cast! {
     v: Content => v.to::<Self>().cloned().ok_or("expected citation")?,
 }
 
-/// A citation style.
+/// A built-in citation style.
 #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Cast)]
-pub enum CitationStyle {
+pub enum CitationStyleBuiltIn {
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/bibliography.rs:402]:
```suggestion
pub enum BuiltInCitationStyle {
```

File: crates/typst-library/src/meta/bibliography.rs [hunk 4]
```diff
@@ -412,16 +418,46 @@ pub enum CitationStyle {
...
+    fn is_short(&self) -> bool {
+        use CitationStyleBuiltIn::*;
+        matches!(self, Self::BuiltIn(Numerical | Alphanumerical | Keys))
     }
 }
 
+/// The result of a cite call, which can either immediately be content or a function call to later
+/// resolve into content.
+#[derive(Clone)]
+enum CiteContentOrFunc {
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/bibliography.rs:450]:
```suggestion
enum ContentOrFunc {
```

File: crates/typst-library/src/meta/bibliography.rs [hunk 5]
```diff
@@ -412,16 +418,46 @@ pub enum CitationStyle {
...
 impl CitationStyle {
-    fn is_short(self) -> bool {
-        matches!(self, Self::Numerical | Self::Alphanumerical | Self::Keys)
+    fn is_short(&self) -> bool {
+        use CitationStyleBuiltIn::*;
+        matches!(self, Self::BuiltIn(Numerical | Alphanumerical | Keys))
     }
 }
 
+/// The result of a cite call, which can either immediately be content or a function call to later
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/bibliography.rs:447]: could you wrap comments at 80 cols?

File: crates/typst-library/src/meta/bibliography.rs [hunk 6]
```diff
@@ -445,6 +481,144 @@ impl Works {
     }
 }
 
+/// Converts a hayagriva title into a typst dictionary.
+fn title_to_dict(title: &hayagriva::types::Title) -> Dict {
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/bibliography.rs:485]: could you move all the conversion helpers to the bottom of the file?

File: crates/typst-library/src/meta/bibliography.rs [hunk 7]
```diff
@@ -496,69 +670,89 @@ fn create(bibliography: BibliographyElem, citations: Vec<CiteElem>) -> Arc<Works
...
+                            Box::new(style::AuthorTitle::new())
+                        }
+                        CitationStyleBuiltIn::Keys => Box::new(style::Keys::new()),
+                    }),
+
+                    CitationStyle::Func(_) => None,
+                };
+            }
+
+            let content = if let Some(citation_style) = citation_style.as_mut() {
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/bibliography.rs:698]: An early return for the `None` case could reduce nesting and increase clarity.

**Comment by laurmaedje** [comment]
Closing this as hayagriva has changed a lot and we're planning to support something like alongside the CSL rework.

**Pull request closed without merging**